### DISCUSSION TITLE
Bundling node binaries in Parcel V2

### DISCUSSION THREAD

**Discussion by quarties** [parcel-bundler/parcel#6178] [Q&A]
Hey, I need to add a node binary to the bundle (it's used by one of the bundled modules). With Parcel V1 I used to do it with this code:

```javascript
const { Asset } = require('parcel-bundler');
const path = require('path');
const URL = require('url');

class NodeAsset extends Asset {
    
    load() {}

    generate() {
        const pathToAsset = this.urlJoin(
            this.options.publicURL,
            this.generateBundleName()
        );
    
        return [
            {
                type: 'js',
                value: `module.exports=eval('require(${JSON.stringify(`.${pathToAsset}`)})');`
            }
        ];
    }

    urlJoin(publicURL, assetPath) {
        const url = URL.parse(publicURL, false, true);
        const assetUrl = URL.parse(assetPath);
        url.pathname = path.posix.join(url.pathname, assetUrl.pathname);
        url.search = assetUrl.search;
        url.hash = assetUrl.hash;

        return URL.format(url);
    }
}

module.exports = NodeAsset;
```

I tried to migrate this transformer to Parcel V2. I was able to add `.node` binary as an asset to the bundle with `isIsolated` option, but I wasn't able to do the `eval('require` trick to make it work. Do you have any ideas how to approach this problem?

**Comment by mischnic** [comment]
Does
```json
{
  "extends": "@parcel/config-default",
  "transformers": {
    "*.node": ["@parcel/transformer-raw"]
  }
}
```

work? You might not even need a custom transformer?

**Reply by quarties** [reply]
I started with this transformer and later tried to modify it. My custom node transformer from parcel v1 was based on the old RawTransformer from V1. As I said - it does add the `.node` binary to the bundle, but it's not properly imported.

In dist the `.node` binary dependency definition looks similar to this:

`module.exports = require(:pathToNodeBianryInBundle:);`

But without eval it won't work, so it should look more like this:

`module.exports = eval('require(:pathToNodeBianryInBundle:)');`

**Reply by mischnic** [reply]
I see, you just get the path with the raw transformer.

This should ideally be fixed in Parcel itself, I don't see why this should require a third party plugin. The fix would be similar to https://github.com/parcel-bundler/parcel/blob/3baf990a686d1c1dd0c5f4360296de890cc00c47/packages/runtimes/js/src/JSRuntime.js#L172-L183
(the bundle type would be `node` in this case).

**Reply by mischnic** [reply]
Related issue: https://github.com/parcel-bundler/parcel/issues/2493
